src/server.ts

```diff
@@ -0,0 +1,108 @@
+/// <reference path="../node_modules/vscode/thenable.d.ts" />
+
+import {
+	IConnection,
+} from 'vscode-languageserver';
+
+import * as net from 'net';
+import * as cluster from 'cluster';
+
+import { newConnection, registerLanguageHandler } from './connection';
+import { registerMasterHandler } from './master-connection';
+import { LanguageHandler } from './lang-handler';
+import * as fs from './fs';
+
+const workersReady = new Map<string, Promise<void>>();
+
+function randomNWorkers(n: number): string[] {
+	const unselected = Array.from(workersReady.keys());
+	let numUnselected = unselected.length;
+	const selected: string[] = [];
+	for (let i = 0; i < n; i++) {
+		const s = Math.floor(Math.random() * numUnselected);
+		selected.push(unselected[s]);
+		const a = unselected[numUnselected - 1], b = unselected[s];
+		unselected[numUnselected - 1] = b, unselected[s] = a;
+		numUnselected--;
+	}
+	return selected;
+}
+
+async function rewriteConsole() {
+	const consoleErr = console.error;
+	console.error = function () {
+		if (cluster.isMaster) {
+			consoleErr(`[mstr]`, ...arguments);
+		} else {
+			consoleErr(`[wkr${cluster.worker.id}]`, ...arguments);
+		}
+	}
+}
+
+/**
+ * serve starts a singleton language server instance that uses a
+ * cluster of worker processes to achieve some semblance of
+ * parallelism.
+ */
+export async function serve(clusterSize: number, lspPort: number, strict: boolean, newLangHandler: () => LanguageHandler): Promise<void> {
+	rewriteConsole();
+
+	if (cluster.isMaster) {
+		console.error(`spawning ${clusterSize} workers`)
+		for (let i = 0; i < clusterSize; ++i) {
+			const worker = cluster.fork().on('disconnect', () => {
+				console.error(`worker ${worker.process.pid} disconnect`)
+			});
+
+			workersReady.set(worker.id, new Promise<void>((resolve, reject) => {
+				worker.on('listening', resolve);
+			}));
+		}
+
+		cluster.on('exit', (worker, code, signal) => {
+			const reason = code === null ? signal : code;
+			console.error(`worker ${worker.process.pid} exit (${reason})`);
+		});
+
+		var server = net.createServer(async (socket) => {
+			const connection = newConnection(socket, socket);
+
+			// Create connections to two worker servers
+			const workerIds = randomNWorkers(2);
```

Comment:
My understanding is that we are using 2 workers from N where N is a number of CPUs. 
- Why not all of them?
- If we want to stick to 2 why create N workers to peek random 2 from them?
- If we want to stick to 2 we should ensure we have at least two CPUs so two workers can be started
- If we want to stick to 2 there is no sense to use cluster size command line parameter